cookbooks/fb_networkmanager/libraries/resource.rb

```diff
@@ -0,0 +1,92 @@
+#
+# Cookbook:: fb_networkmanager
+# Recipe:: default
+#
+# Copyright (c) 2020-present, Vicarious, Inc.
+# Copyright (c) 2020-present, Facebook, Inc.
+# All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module FB
+  class Networkmanager
+    module Resource
+      NM_CONN_DIR = '/etc/NetworkManager/system-connections'.freeze
+      PREFIX = 'fb_networkmanager'.freeze
+
+      def conf_path(name, normalize = true)
+        if normalize
+          n = "#{PREFIX}_#{name.downcase.gsub(' ', '_')}"
+        else
+          n = name
+        end
+        ::File.join(NM_CONN_DIR, n)
+      end
+
+      def determine_files(name, config)
+        cfile = conf_path(name)
+        migratefile = fromfile = nil
+        if config['_migrate_from']
+          migratefile = conf_path(config['_migrate_from'], false)
+          # Once we've created a file, we should no longer pay attention to the
+          # old file
+          if ::File.exist?(cfile)
+            fromfile = cfile
+          elsif ::File.exist?(migratefile)
+            Chef::Log.info(
+              "#{PREFIX}: Building #{name} based on #{migratefile}",
+            )
+            fromfile = migratefile
+          else
+            # Neither the new file nor the old file exist...
+            Chef::Log.warn(
+              "#{PREFIX}: We were asked to migrate from #{migratefile} but " +
+              'that does not exist, so making a fresh config',
+            )
+            fromfile = cfile
+          end
+          config.delete('_migrate_from')
+        else
+          fromfile = cfile
+        end
+
+        return {
+          'config' => cfile,
+          'from' => fromfile,
+          'migrate' => migratefile,
+        }
+      end
+
+      def allowed_connections(node)
+        node['fb_networkmanager']['system_connections'].keys.map do |x|
+          "#{PREFIX}_#{x.downcase.gsub(' ', '_')}"
+        end
+      end
+
+      # the magic here is all DeepMerge and testing this is arguably a bit
+      # silly, but since this is the most important part of the whole cookbook,
+      # we add a test just to make sure no one ever reverses the arguments or
+      # something weird
+      def generate_config_hashes(from_file, desired_config)
+        current = {}
+        if ::File.exist?(from_file)
+          current = IniParse.parse(::File.read(from_file)).to_hash
```

Comment:
I dropped trailing semi-colons and it worked fine...

Reply:
I checked with NetworkManager 1.20 and 1.22, when dropping trailing semicolons I had to restart the NetworkManager service (not just reload connections) and then it worked, but as soon as anything modified it (e.g. NetworkManager updates `seen-bssids`) trailing semicolons were back.

Reply:
This is interesting stuff. I'll dig into that, thanks Joel.

Reply:
OK so, @jgoguen  I cannot repro your experience:

Here's my home network, I dropped the trailing `;` from `seen-bssids`:
```
root@ldx-vulturus:/etc/NetworkManager/system-connections# grep ';' philnet2
seen-bssids=19:F8:29:65:26:59;19:F8:29:64:26:59
```

Note I have no DNS search domains:
```
root@ldx-vulturus:/etc/NetworkManager/system-connections# grep search philnet2
dns-search=
dns-search=
```

Which matches what NM knows:

```
root@ldx-vulturus:/etc/NetworkManager/system-connections# nmcli c show philnet2 | egrep 'bssids|dns-search'
802-11-wireless.seen-bssids:            19:F8:29:65:26:59,10:F8:29:64:26:59
ipv4.dns-search:                        --
ipv6.dns-search:                        --
```

So now let me add a DNS search domain:

```
root@ldx-vulturus:/etc/NetworkManager/system-connections# grep search philnet2
dns-search=test.test
dns-search=
```

And before I reload, nm doesn't know about the change, as expected, but then I reload and nmcli knows about it:

```
root@ldx-vulturus:/etc/NetworkManager/system-connections# nmcli c show philnet2 | egrep 'dns-search'
ipv4.dns-search:                        --
ipv6.dns-search:                        --
root@ldx-vulturus:/etc/NetworkManager/system-connections# nmcli c reload
root@ldx-vulturus:/etc/NetworkManager/system-connections# nmcli c show philnet2 | egrep 'dns-search'
ipv4.dns-search:                        test.test
ipv6.dns-search:                        --
root@ldx-vulturus:/etc/NetworkManager/system-connections#
```

While it's possible they introduced some bug later, it's also possible that there was some other thing going on in your testing. Can you try the above experiment?

Reply:
I really hate NetworkManager sometimes. I've rebooted since last night, maybe there was something screwed up on my end, but I can't reproduce this again either with my original test or repeating exactly yours.